File: user_app/lib/features/models/address/address_model.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:Quickzon/core/utils/enums/enums.dart';

part 'address_model.g.dart';

@JsonSerializable(
  explicitToJson: true,
  includeIfNull: false, // 🔥 IMPORTANT
)
class AddressType {
  @JsonKey(name: '_id')
  final String? id;

  final String userId;
  final String label;
  final AddressLabelType type;

  @JsonKey(includeIfNull: false)
  final String? customLabel;

  final String fullName;
  final String phone;

  @JsonKey(includeIfNull: false)
  final String? alternatePhone;

  final String addressLine1;

  @JsonKey(includeIfNull: false)
  final String? addressLine2;

  @JsonKey(includeIfNull: false)
  final String? googleLocation;

  @JsonKey(includeIfNull: false)
  final String? landmark;

  final String city;
  final String state;
  final String pincode;
  final String country;

  @JsonKey(includeIfNull: false)
  final String? formattedAddress;

  @JsonKey(includeIfNull: false)
  final GeoLocation? location;

  @JsonKey(includeIfNull: false)
  final int? placeId;

  @JsonKey(includeIfNull: false)
  final String? osmType;

  @JsonKey(includeIfNull: false)
  final String? addressType;

  final bool isDefault;
  final bool isActive;

  @JsonKey(includeIfNull: false)
  final DateTime? createdAt;

  @JsonKey(includeIfNull: false)
  final DateTime? updatedAt;

  AddressType({
    this.id,
    required this.userId,
    required this.label,
    required this.type,
    this.customLabel,
    required this.fullName,
    required this.phone,
    this.alternatePhone,
    required this.addressLine1,
    this.addressLine2,
    this.googleLocation,
    this.landmark,
    required this.city,
    required this.state,
    required this.pincode,
    required this.country,
    this.formattedAddress,
    this.location,
    this.placeId,
    this.osmType,
    this.addressType,
    this.isDefault = false,
    this.isActive = true,
    this.createdAt,
    this.updatedAt,
  });

  factory AddressType.fromJson(Map<String, dynamic> json) =>
      _$AddressTypeFromJson(json);

  Map<String, dynamic> toJson() => _$AddressTypeToJson(this);
}

@JsonSerializable(includeIfNull: false)
class GeoLocation {
  final String type;
  final List<double> coordinates;

  GeoLocation({
    required this.type,
    required this.coordinates,
  });

  factory GeoLocation.fromJson(Map<String, dynamic> json) =>
      _$GeoLocationFromJson(json);

  Map<String, dynamic> toJson() => _$GeoLocationToJson(this);
}

File: server/src/models/address/address.model.ts
import mongoose from "mongoose";
import {
  IAddressDocument,
  AddressModelType,
  AddressLabelType,
} from "../../lib/types/index";
import { addressUtils } from "./address.utils";

const AddressSchema = new mongoose.Schema<IAddressDocument, AddressModelType>(
  {
    userId: {type: mongoose.Schema.Types.ObjectId,ref: "User",required: true,index: true,},
    label: { type: String, required: true, trim: true, maxlength: 50 },
    type: {type: String,enum: Object.values(AddressLabelType),required: true,default: AddressLabelType.HOME,},
    customLabel: { type: String, trim: true, maxlength: 50, default: null },
    fullName: { type: String, required: true, trim: true, maxlength: 100 },
    phone: {type: String,required: true,trim: true,match: [/^[0-9]{10}$/, "Please enter a valid 10-digit phone number"],},
    alternatePhone: {type: String,trim: true,match: [/^[0-9]{10}$/, "Please enter a valid 10-digit phone number"],},
    addressLine1: { type: String, required: true, trim: true, maxlength: 200 },
    addressLine2: { type: String, trim: true, maxlength: 200 },
    googleLocation: { type: String, trim: true, maxlength: 200 },
    landmark: { type: String, trim: true, maxlength: 100 },
    city: { type: String, required: true, trim: true, maxlength: 100 },
    state: { type: String, required: true, trim: true, maxlength: 100 },
    pincode: {type: String,required: true,trim: true,match: [/^[0-9]{6}$/, "Please enter a valid 6-digit pincode"],},
    country: { type: String, required: true, default: "India", trim: true },
    formattedAddress: { type: String, trim: true },
    // location: {type: {  type: String,  enum: ["Point"],  default: "Point",},coordinates: {  type: [Number],  required: true,  validate: {    validator: (arr: number[]) => arr.length === 2,    message: "Coordinates must be [longitude, latitude]",  },},},
    placeId: { type: String, trim: true },
    osmType: { type: String, trim: true },
    addressType: { type: String, trim: true },
    isDefault: { type: Boolean, default: false },
    isActive: { type: Boolean, default: true },
  },
  { timestamps: true }
);

AddressSchema.index({ userId: 1, isActive: 1 });
AddressSchema.index({ userId: 1, isDefault: 1 });
AddressSchema.index({ location: "2dsphere" });

AddressSchema.statics = addressUtils.statics;

export const Address = mongoose.model<IAddressDocument, AddressModelType>(
  "Address",
  AddressSchema
);
